### ISSUE TITLE
Improve keybindings on spacemacs buffer (home buffer)

### ISSUE THREAD

**Issue by rommeswi** [syl20bnr/spacemacs#14230]
#### Description :octocat:
The keyboard shortcuts for "recent", "projects", etc. are inconvenient. Navigating the spacemacs buffer is not streamlined.

#### Reproduction guide :beetle:
- Start Emacs.
- Press 'r'.
- Cursor jumps to first recent file.
- Press 'r' again.

*Observed behaviour:* :eyes: :broken_heart:
Cursor does not move to the second recent file after pressing 'r' the second time.

*Expected behaviour:* :heart: :smile:
Pressing r should consecutively cycle through all recent files.
Pressing R should cycle backwards through all recent files.

Similar behavior is expected for projects, agenda files, etc.

#### System Info :computer:
- OS: gnu/linux
- Emacs: 26.3
- Spacemacs: 0.300.0
- Spacemacs branch: develop (rev. 62979df31)
- Graphic display: t
- Distribution: spacemacs
- Editing style: vim
- Completion: ivy
- Layers:
```elisp
(ivy emacs-lisp
     (evil-snipe :variables evil-snipe-enable-alternate-f-and-t-behaviors t)
     git bibtex latex pdf mu4e multiple-cursors org spell-checking treemacs)
```
- System configuration features: XPM JPEG TIFF GIF PNG RSVG IMAGEMAGICK SOUND GPM DBUS GSETTINGS GLIB NOTIFY ACL LIBSELINUX GNUTLS LIBXML2 FREETYPE M17N_FLT LIBOTF XFT ZLIB TOOLKIT_SCROLL_BARS GTK3 X11 XDBE XIM THREADS LIBSYSTEMD LCMS2

**Comment by blablablerg** [comment]
Ha I didn't know this, nice! But in evil mode j/k works to go up and down. r goes to the first line. Seems fine to me this way?

**Comment by rommeswi** [comment]
I know about the j/k shortcut (Tab also works). I guess it has become just a pet peeve of me to get annoyed every time I have to press another button after "r" to jump to the second line when (just as with searches and "n"/"N") I want to press "r" a second time. In a perfect world, I would even like to cycle through more than just the number of documents displayed. So if I've set my config to display 3 documents, pressing "r" a fourth time would scroll the whole list upwards and display the next document...and....I guess I'm dreaming now....

**Comment by duianto** [comment]
Here's a possible solution to cycle through the lists.
(have only tested them with Recent Files and Projects)

Add this code block to the `.spacemacs`, `dotspacemacs/user-config` section.
```elisp
(defun spacemacs/goto-startup-list-header (list-name)
  (unless (search-forward list-name (point-max) 'noerror)
    (search-backward list-name (point-min) 'noerror)))
(defun spacemacs/goto-startup-list-first-item (list-name)
  (spacemacs/goto-startup-list-header list-name)
  (forward-line 1)
  (back-to-indentation))
(defun spacemacs/get-startup-list-header-line (list-name)
  (save-excursion
    (spacemacs/goto-startup-list-header list-name)
    (line-number-at-pos)))
(defun spacemacs/get-startup-list-name (list-name)
  "Given a Spacemacs buffer startup list title.
Return the startup lists name:
Given          Returns
Recent Files:  recents
Bookmarks:     bookmarks
Projects:      projects
Agenda:        agenda
ToDo:          todos"
  (let ((found-space (string-match " " list-name)))
    (concat (downcase
             (substring list-name 0 (or found-space (string-match ":" list-name))))
            (when (or found-space (string= list-name "ToDo:")) "s"))))
(defun spacemacs/get-nr-of-list-items (list-name)
  "Given a startup list name.
Return the number of configured list items."
  (let (nr)
    (dolist (item dotspacemacs-startup-lists)
     (when (string= (car item)
                    (spacemacs/get-startup-list-name list-name))
       (setq nr (cdr item))))
    nr))
(defun spacemacs/on-a-startup-list-item (list-name)
  (and (> (line-number-at-pos)
          (spacemacs/get-startup-list-header-line list-name))
       (<= (line-number-at-pos)
           (+ (spacemacs/get-startup-list-header-line list-name)
              (spacemacs/get-nr-of-list-items list-name)))))
(defun spacemacs/on-last-startup-list-item (list-name)
  (= (line-number-at-pos)
     (+ (spacemacs/get-startup-list-header-line list-name)
        (spacemacs/get-nr-of-list-items list-name))))
(defun spacemacs/before-first-startup-list-item (list-name)
  (<= (line-number-at-pos)
      (spacemacs/get-startup-list-header-line list-name)))
(defun spacemacs/after-last-startup-list-item (list-name)
  (> (line-number-at-pos)
     (+ (spacemacs/get-startup-list-header-line list-name)
        (spacemacs/get-nr-of-list-items list-name))))
(defun spacemacs/on-an-empty-line ()
  (string= (buffer-substring (line-beginning-position)
                             (line-end-position)) ""))
(defun spacemacs/empty-line-after-startup-list-item ()
  "Check if the next line is empty,
while the line still is within the
number of configured startup list items.
This was used in the Projects list,
which was configured to show 7 projects,
but only 3 projects had been added."
  (save-excursion
    (forward-line 1)
    (spacemacs/on-an-empty-line)))
(defun spacemacs/cycle-startup-list-items (list-name)
  (cond ((or (spacemacs/on-last-startup-list-item list-name)
             (spacemacs/before-first-startup-list-item list-name)
             (spacemacs/after-last-startup-list-item list-name)
             (spacemacs/empty-line-after-startup-list-item))
         (spacemacs/goto-startup-list-first-item list-name))
        ((spacemacs/on-a-startup-list-item list-name)
         (forward-line 1)
         (back-to-indentation))))
(defun spacemacs/home-buffer-cycle-recent-files ()
  (interactive)
  (spacemacs/cycle-startup-list-items "Recent Files:"))
(defun spacemacs/home-buffer-cycle-projects ()
  (interactive)
  (spacemacs/cycle-startup-list-items "Projects:"))
(global-set-key (kbd "<f8>") 'spacemacs/home-buffer-cycle-recent-files)
(global-set-key (kbd "<f9>") 'spacemacs/home-buffer-cycle-projects)
(define-key spacemacs-buffer-mode-map (kbd "R") 'spacemacs/home-buffer-cycle-recent-files)
(define-key spacemacs-buffer-mode-map (kbd "P") 'spacemacs/home-buffer-cycle-projects)
;; the following rebindings of the lower case keys: r and p
;; don't work on startup. The default commands that Spacemacs defined are still bound.
;; (define-key spacemacs-buffer-mode-map (kbd "r") 'spacemacs/home-buffer-cycle-recent-files)
;; (define-key spacemacs-buffer-mode-map (kbd "p") 'spacemacs/home-buffer-cycle-projects)
```

**Comment by lebensterben** [comment]
@duianto 
I have a much simpler (and less powerful) hack.

By default `C-x ]` and `C-x [` jumps to the start of next and previous page break respectively.
It matches the variable `page-delimeter`, whose default value is a line starts with form feed `^\f`.

We can simply wrap the call to `forward-page` and `backward-page` in a `let` form and set `page-delimeter` to unicode `^` plus unicode zero width space. And prepend this zero width space to each startup list title.

The zero width space is normally invisible, **unless** in a terminal that doesn't support unicode input such as tty. This is the only downside.

---

If you want to get fancy, our wrapper function can accept an argument, and set `page-delimeter` accordingly.
For example, the i-th element in the startup lists would have i zero width space prepended. And we can have a binding `g N` that goes to N-th list.

**Comment by duianto** [comment]
That's a clever solution.

I just remembered this related feature request.
Is it able to show the index of the recent files/projects in spacemacs home page? #12519

It seems even simpler to press a number to jump to a file, instead of pressing the same key multiple times.

**Comment by duianto** [comment]
A PR has been opened with a possible solution:
Add home buffer recent files numbers #14326

**Comment by Invertisment** [comment]
I was using evil `j/k` with repeat before the upstream changed the file opening to number-based. Like this:
`20j`

This would go 20 lines down. Then `o` would show the openable files. But now I can't do the jump using the number (except if I use Emacs style of specifying number arg for next command (and in that one `2 0 j` becomes `C-2 C-0 j`)).

There are no paragraphs so `C-}` also doesn't do whole much. I'd like to use this as there would be less keystrokes. But I'm not sure if home buffer has regular paragraphs.

Also there is this text too and it changes so I don't even know how many lines to jump down:

```
                                                         ╭─ Important Notes (Release 0.300.x) ────────────────────────────────────────────────╮
                                                         │                                                                                    │
                                                         │         ╭───────────────────────────────────────────────────────────╮              │
                                                         │                                                                                    │
                                                         │                         VERSION 0.300 IS ALMOST OUT!                               │
                                                         │                                                                                    │
                                                         │         ╰───────────────────────────────────────────────────────────╯              │
                                                         │                                                                                    │
                                                         │  The long awaited feature to have stable ELPA packages is now available in the     │
                                                         │  develop branch and we need to test it out before we release the 0.300 version.    │
                                                         │                                                                                    │
                                                         │  The stable ELPA repository is a Git repository, the official one is called        │
                                                         │  Spacelpa and it is hosted on GitHub here:                                         │
                                                         │                                                                                    │
                                                         │      https://github.com/syl20bnr/spacelpa                                          │
                                                         │                                                                                    │
                                                         │  In the end only the master branch will use the stable ELPA repository and the     │
                                                         │  develop branch will continue to use bleeding edge packages from MELPA             │
                                                         │  essentially.                                                                      │
                                                         │                                                                                    │
                                                         │  Unfortunately this new feature requires at least Emacs 25.1 to work correctly as  │
                                                         │  we need the archive priority feature of package.el that shipped with this         │
                                                         │  version of Emacs.                                                                 │
                                                         │                                                                                    │
                                                         │  The ELPA repository configuration is set in a new immutable file called .lock     │
                                                         │  that sits at the root of the Spacemacs git repository.                            │
                                                         │                                                                                    │
                                                         │  Spacemacs downloads the whole ELPA stable repository locally so it means that     │
                                                         │  once it is installed you don't need an Internet connection anymore to install     │
                                                         │  any packages covered by Spacemacs layers!                                         │
                                                         │                                                                                    │
                                                         │  Default installation location of the ELPA stable repository is in:                │
                                                         │                                                                                    │
                                                         │      ~/.emacs.d/.cache/stable-elpa                                                 │
                                                         │                                                                                    │
                                                         │  If you want to disable the ELPA stable repository put this in your dotfile in     │
                                                         │  the user-init function:                                                           │
                                                         │                                                                                    │
                                                         │      (setq configuration-layer-elpa-archives '(("melpa" . "melpa.org/packages/")   │
                                                         │            ("org" . "orgmode.org/elpa/") ("gnu" . "elpa.gnu.org/packages/")))      │
                                                         │                                                                                    │
                                                         ╰─ Update your dotfile (SPC f e D) and packages after every update ──────────────────╯


```

**Comment by lebensterben** [comment]
@Invertisment 
The release note can be turned off.
Why do you prefer to "jump down" when you can simply type the number before an entry to open it?

**Comment by duianto** [comment]
The open/close state of the notes on the Spacemacs home buffer,
can be toggled by pressing `RET` (return/enter) on the menu buttons:

- `[?]` shows/hides the note: `Quick Help`
- `[Release Notes]` shows/hides the note: `Important Notes`

The notes visibility is remembered when Spacemacs is restarted.

**Comment by Invertisment** [comment]
> @Invertisment
> The release note can be turned off.
> Why do you prefer to "jump down" when you can simply type the number before an entry to open it?

Well... There are no numbers on my screen. :D At least when I don't jump down :D

![Screenshot at 2021-05-08 15-34-31](https://user-images.githubusercontent.com/1641263/117539354-e999c200-b012-11eb-8258-3706a3d045f9.png)

**Comment by lebensterben** [comment]
> Why doesn't the message say that you can turn it off if its intention is to be seen once

Isn't this obvious?

If you are unclear about something
- Ask about it
- Read the source code

Don't complain about something unrelated.

**Comment by duianto** [comment]
I'll open a PR to add a `[Close note]` button to the bottom right of the notes, to make it more visible how they can be closed.

Quick Help
![emacs_2021-05-08_17-35-04](https://user-images.githubusercontent.com/13420573/117545119-c24bf080-b024-11eb-8dc2-9ed6a5ea69f7.png)

Release Notes
![emacs_2021-05-08_17-35-31](https://user-images.githubusercontent.com/13420573/117545116-bfe99680-b024-11eb-86a3-406fc426ca12.png)

**Comment by duianto** [comment]
>The thing is that the numbers still work even though I don't see what they mark if I'm at the top of the buffer. So if I click any number a random buffer may open up. :D

That is unexpected behavior. Thanks.
Although it would probably not be a good idea to be able to jump by line with (nr)j or (nr)k,
when the numbers aren't visible.

Because when one jumps down and the numbers become visible, 
then the behavior of the number keys will change 
and they will open the target file/directory instead.

One solution might be to allow for disabling the numbers.
For anyone that only want to open them by navigating there manually and pressing `RET`
or mouse clicking to open a file/directory.

>But why isn't then it turned off by default?

The purpose of the note is to provide information about important changes.

>Who needs to see it then?
If an important change has occurred, then most (all?) Spacemacs users probably would want to know about it.

>What is the use of it if its purpose is to be disabled?
See above.

>Why doesn't the message say that you can turn it off if its intention is to be seen once?
Your right that would be helpful.
An attempt at fixing this is pending in the PR mentioned above:
[home] Add button: Close note #14749

>And the numbers look like this on my screen (I wanted to have a little bit more recent projects):
Could you clarify what isn't working?

**Comment by Invertisment** [comment]
> > The thing is that the numbers still work even though I don't see what they mark if I'm at the top of the buffer. So if I click any number a random buffer may open up. :D
> 
> That is unexpected behavior. Thanks.
> #### Update
> 
> PR merged: [home] Only jump to visible nr lines #14750
> 
> Although it would probably not be a good idea to be able to jump by line with (nr)j or (nr)k,
> when the numbers aren't visible.
> 
> Because when one jumps down and the numbers become visible,
> then the behavior of the number keys will change
> and they will open the target file/directory instead.
> 
> One solution might be to allow for disabling the numbers.
> For anyone that only want to open them by navigating there manually and pressing `RET`
> or mouse clicking to open a file/directory.
> #### Update
> 
> PR merged: [home] New var dotspacemacs-show-startup-list-numbers #14751
> 
> The new variable can be added to your `.spacemacs`
> by comparing your `.spacemacs` with the `.spacemacs.template` using ediff,
> by pressing: `SPC f e D`
> 
> > But why isn't then it turned off by default?
> 
> The purpose of the note is to provide information about important changes.
> 
> > Who needs to see it then?
> 
> If an important change has occurred, then most (all?) Spacemacs users probably would want to know about it.
> 
> > What is the use of it if its purpose is to be disabled?
> 
> See above.
> 
> > Why doesn't the message say that you can turn it off if its intention is to be seen once?
> 
> Your right that would be helpful.
> 
> A `Close note` button has been added below the notes:
> [home] Add button: Close note #14749
> 
> > And the numbers look like this on my screen (I wanted to have a little bit more recent projects):
> 
> Could you clarify what isn't working?

I asked problemic questions. Some of them were simply drama and additional details. I'll try to not do that again.
